controllers/clusterresources/opensearchegressrules_controller.go

```diff
@@ -0,0 +1,196 @@
+//+kubebuilder:rbac:groups=clusterresources.instaclustr.com,resources=opensearchegressrules/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=clusterresources.instaclustr.com,resources=opensearchegressrules/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the OpenSearchEgressRules object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.14.4/pkg/reconcile
+func (r *OpenSearchEgressRulesReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	l := log.FromContext(ctx)
+
+	rule := &clusterresourcesv1beta1.OpenSearchEgressRules{}
+	err := r.Client.Get(ctx, types.NamespacedName{
+		Namespace: req.Namespace,
+		Name:      req.Name,
+	}, rule)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			return models.ExitReconcile, nil
+		}
+
+		l.Error(err, "Unable to fetch OpenSearch Egress Rules resource")
+
+		return models.ReconcileRequeue, err
+	}
+
+	// Handle resource deletion
+	if rule.DeletionTimestamp != nil {
+		err = r.handleDelete(ctx, l, rule)
+		if err != nil {
+			return models.ReconcileRequeue, err
+		}
+
+		return models.ExitReconcile, nil
+	}
+
+	// Handle resource creation
+	if rule.Status.ID == "" {
+		err = r.handleCreate(ctx, l, rule)
+		if err != nil {
+			return models.ReconcileRequeue, nil
+		}
+
+		return models.ExitReconcile, nil
+	}
+
+	return models.ExitReconcile, nil
+}
+
+func (r *OpenSearchEgressRulesReconciler) handleCreate(ctx context.Context, l logr.Logger, rule *clusterresourcesv1beta1.OpenSearchEgressRules) error {
+	b, err := r.API.CreateOpenSearchEgressRules(rule.Spec.ClusterID, rule.Spec)
+	if err != nil {
+		l.Error(err, "failed to create OpenSearch Egress Rule resource on Instaclustr")
+		r.EventRecorder.Eventf(rule, models.Warning, models.CreationFailed,
+			"Failed to create OpenSearch Egress Rule on Instaclustr. Reason: %v", err,
+		)
+
+		return err
+	}
+
+	patch := rule.NewPatch()
+	err = json.Unmarshal(b, &rule.Status)
+	if err != nil {
+		l.Error(err, "failed to parse OpenSearch Egress Rule resource response from Instaclustr")
+		r.EventRecorder.Eventf(rule, models.Warning, models.ConvertionFailed,
+			"Failed to parse OpenSearch Egress Rule response from Instaclustr. Reason: %v", err,
+		)
+
+		return err
+	}
+
+	// This hack was added because instaclustr api returns null value egress rule id, but it should be equal to pattern {clusterId}~{source}~{bindingId}
+	// This code could be removed after instaclustr fix bugs, to make code cleaner
```

Comment:
// This hack was added because the Instaclustr API returns an egress rule id with a null value, but it should be equal to the pattern {clusterId}~{source}~{bindingId}
// This code could be removed after the Instaclustr API team fixes this bug